tmail-backend/jmap/extensions/src/main/scala/com/linagora/tmail/james/jmap/publicAsset/PublicAssetRepository.scala

```diff
@@ -0,0 +1,90 @@
+
+  def revoke(username: Username, id: PublicAssetId): Publisher[Void]
+
+  def revoke(username: Username): Publisher[Void]
+
+  def get(username: Username, ids: Set[PublicAssetId]): Publisher[PublicAsset]
+
+  def list(username: Username): Publisher[PublicAsset]
+
+}
+
+class PublicAssetService @Inject()(uploadRepository: UploadRepository,
+                                   @Named("jmap") jmapConfiguration: Configuration) {
+  private val publicURIPrefix: String = jmapConfiguration.getString("publicAsset.publicURIPrefix", "")
+
+  def retrieveMetaData(id: UploadId, username: Username): SMono[(Size, ImageContentType)] =
+    SFlux(uploadRepository.listUploads(username))
+      .filter(_.uploadId.equals(id))
+      .next()
+      .map(uploadMetaData => (uploadMetaData.size, ImageContentType.from(uploadMetaData.contentType)))
+
+  def generatePublicURI(publicAssetId: PublicAssetId, accountId: AccountId): PublicURI =
+    PublicURI.from(publicAssetId, accountId, publicURIPrefix)
+
+  def generatePublicAsset(uploadId: UploadId, username: Username): SMono[PublicAsset] =
+    SFlux(uploadRepository.listUploads(username))
+      .filter(_.uploadId.equals(uploadId))
+      .next()
+      .map(uploadMetadata => {
+        val publicAssetId: PublicAssetId = PublicAssetIdFactory.generate()
+        val publicURI: PublicURI = generatePublicURI(publicAssetId, AccountId.from(username).fold(e => throw e, identity))
+        PublicAsset(
+          id = publicAssetId,
+          publicURI = publicURI,
+          size = uploadMetadata.size,
+          contentType = ImageContentType.from(uploadMetadata.contentType),
+          blobId = uploadMetadata.blobId)
+      })
+}
+
+class MemoryPublicAssetRepository()(val publicAssetService: PublicAssetService) extends PublicAssetRepository {
+  private val tableStore: Table[Username, PublicAssetId, PublicAsset] = Tables.synchronizedTable(HashBasedTable.create())
+
+  override def create(username: Username, creationRequest: PublicAssetCreationRequest): SMono[PublicAsset] =
+    publicAssetService.generatePublicAsset(creationRequest.uploadId, username)
+      .doOnNext(publicAsset => tableStore.put(username, publicAsset.id, publicAsset))
+
+  override def update(username: Username, id: PublicAssetId, identityIds: Set[IdentityId]): SMono[Void] =
+    SMono.fromCallable(() => tableStore.get(username, id))
+      .switchIfEmpty(SMono.error(PublicAssetNotFoundException(id)))
+      .map { publicAsset =>
+        val updatedPublicAsset = publicAsset.copy(identityIds = identityIds.toSeq)
+        tableStore.put(username, id, updatedPublicAsset)
+      }.`then`(SMono.empty)
+
+  override def revoke(username: Username, id: PublicAssetId): SMono[Void] =
+    SMono.fromCallable(() => tableStore.remove(username, id))
+      .`then`(SMono.empty)
+
+  override def revoke(username: Username): SMono[Void] =
+    SMono.fromCallable(() => tableStore.row(username).clear())
+      .`then`(SMono.empty)
```

Comment:
Following your current way, we should delete the blob too IMO.

Reply:
I will update with the way that deletes blob in JmapMethod

Reply:
We did agree to do blob operations (copy in create, delete in destroy) in the jmap level and not repository :)